cobigen/cobigen-core/src/main/java/com/devonfw/cobigen/impl/config/upgrade/TemplateSetUpgrader.java

```diff
@@ -0,0 +1,312 @@
+				.field("retrieveObjectsRecursively:{isRetrieveObjectsRecursively|setRetrieveObjectsRecursively(new Boolean(%s))|type=java.lang.Boolean}",
+						"retrieveObjectsRecursively:{isRetrieveObjectsRecursively|setRetrieveObjectsRecursively(new Boolean(%s))|type=java.lang.Boolean}")
+				.byDefault().register();
+		this.mapper = mapperFactory.getMapperFacade();
+	}
+
+
+	/**
+	 * Upgrades the ContextConfiguration from v2.1 to the new structure from v3.0. The
+	 * monolithic pom and context files will be split into multiple files
+	 * corresponding to every template set that will be created.
+	 *
+	 * @param {@link Path} Path to the context.xml that will be upgraded
+	 * @return {@link Map} collection that contains the upgraded v3.0 {@link com.devonfw.cobigen.impl.config.entity.io.v3_0.ContextConfiguration}
+	 * 			as key and a {@link Path} for the new location of the context.xml as value
+	 * @throws Exception
+	 */
+	public Map<com.devonfw.cobigen.impl.config.entity.io.v3_0.ContextConfiguration, Path> upgradeTemplatesToTemplateSets(
+			Path contextLocation) throws Exception {
+		Path context = analyseStructure(contextLocation);
+		ContextConfiguration contextConfiguration = getContextConfiguration(context);
+
+		Path cobigenDir = context;
+		while (!cobigenDir.endsWith(ConfigurationConstants.COBIGEN_CONFIG_FILE)) {
+			cobigenDir = cobigenDir.getParent();
+		}
+
+		Path templateSets = Files.createDirectory(cobigenDir.resolve(ConfigurationConstants.TEMPLATE_SETS_FOLDER));
+		Path adapted = Files.createDirectory(templateSets.resolve(ConfigurationConstants.ADAPTED_FOLDER));
+		Path cobigenTemplates = cobigenDir.resolve(ConfigurationConstants.TEMPLATES_FOLDER)
+				.resolve(ConfigurationConstants.COBIGEN_TEMPLATES);
+		Path templates = cobigenTemplates.resolve(ConfigurationConstants.TEMPLATE_RESOURCE_FOLDER);
+
+		List<com.devonfw.cobigen.impl.config.entity.io.v3_0.ContextConfiguration> contextFiles = splitContext(
+				contextConfiguration);
+		Map<com.devonfw.cobigen.impl.config.entity.io.v3_0.ContextConfiguration, Path> contextMap = new HashMap<>();
+		for (com.devonfw.cobigen.impl.config.entity.io.v3_0.ContextConfiguration cc : contextFiles) {
+			for (com.devonfw.cobigen.impl.config.entity.io.v3_0.Trigger trigger : cc.getTrigger()) {
+				Path triggerFolder = templates.resolve(trigger.getTemplateFolder());
+				Path newTriggerFolder = adapted.resolve(trigger.getTemplateFolder());
+				Path utilsPath = cobigenTemplates.resolve("src/main/java");
+				try {
+					FileUtils.copyDirectory(triggerFolder.toFile(),
+							newTriggerFolder.resolve("src/main/resources").toFile());
+				} catch (Exception e) {
+					LOG.error("Could not copy template Folder with the Error" + e);
+					throw e;
+				}
+				try {
+					FileUtils.copyDirectory(utilsPath.toFile(), newTriggerFolder.resolve("src/main/java").toFile());
+				} catch (Exception e) {
+					LOG.error("Could not copy utlis Folder with the Error" + e);
```

Comment:
Better use: LOG.error("An error occurred while copying the template utilities folder!", e);